elyra/pipeline/component.py

```diff
@@ -0,0 +1,233 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from abc import abstractmethod
+from traitlets.config import LoggingConfigurable
+from typing import List, Optional
+
+
+cardinality = {
+    'min': 0,
+    'max': -1
+}
+
+inputs = {
+    "id": "inPort",
+    "app_data": {
+        "ui_data": {
+            "cardinality": cardinality,
+            "label": "Input Port"
+        }
+    }
+}
+
+outputs = {
+    "id": "outPort",
+    "app_data": {
+        "ui_data": {
+            "cardinality": cardinality,
+            "label": "Output Port"
+        }
+    }
+}
+
+empty_properties = {
+    "current_parameters": {"component_source": "", "runtime_image": "", "component_source_type": ""},
+    "parameters": [{"id": "component_source"}, {"id": "runtime_image"}, {"id": "component_source_type"}],
+    "uihints": {
+        "id": "nodeProperties",
+        "parameter_info": [
+            {
+                "parameter_ref": "component_source",
+                "control": "readonly",
+                "label": {
+                    "default": "Path to Component"
+                },
+                "description": {
+                    "default": "The path to the component specification file.",
+                    "placement": "on_panel"
+                },
+                "data": {
+                    "format": "string"
+                }
+            },
+            {
+                "parameter_ref": "runtime_image",
+                "control": "custom",
+                "custom_control_id": "EnumControl",
+                "label": {
+                    "default": "Runtime Image"
+                },
+                "description": {
+                    "default": "Docker image used as execution environment.",
+                    "placement": "on_panel"
+                },
+                "data": {
+                    "items": [],
+                    "required": True
+                }
+            },
+            {
+                "parameter_ref": "component_source_type",
+                "control": "readonly",
+                "label": {
+                    "default": "Component Source Type"
+                },
+                "data": {
+                    "format": "string"
+                }
+            }
+        ],
+        "group_info": [
+            {
+                "id": "nodeGroupInfo",
+                "type": "panels",
+                "group_info": [
+                    {"id": "component_source", "type": "controls", "parameter_refs": ["component_source"]},
+                    {"id": "runtime_image", "type": "controls", "parameter_refs": ["runtime_image"]},
+                    {"id": "component_source_type", "type": "controls", "parameter_refs": ["component_source_type"]}
+                ]
+            }
+        ]
+    },
+    "resources": {}
+}
+
+
+class Property(object):
```

Comment:
Should this be named `ComponentProperty` or something other than just `Property`?  That seems a bit too generic.